### PR TITLE
[custom ops] Support factory function 

### LINKED ISSUE #129389: [custom_op] factory functions don't work
```py
import torch
from torch import Tensor
from typing import Tuple

@torch.library.custom_op("mylib::bar", mutates_args={}, device_types="cpu")
def bar(device: torch.device) -> Tensor:
    return torch.ones(3)

bar(cpu)
```
the above raises an error.

## Proposal (may need more design)

- If a user registers a device-specific implementation for an operator that accepts no Tensors, then we require the operator to have a "device: torch.device argument"
- We switch on the device argument to select the correct backend to dispatch to.
- In the future, we can allow users to customize this behavior (they probably want to dispatch to more than just device)

cc @ezyang @anijain2305 @chauhang @bdhirsh

### PR DESCRIPTION
**Pull request by yushangdi**
Fixes #129389 

If a user registers a device-specific implementation for an operator that accepts no Tensors, then we require the operator to have a "device: torch.device argument"

We switch on the device argument to select the correct backend to dispatch to.

### CODE REVIEW

File: torch/_library/custom_ops.py [hunk 1]
```diff
@@ -502,6 +532,19 @@ def adinplaceorview_impl(keyset, *args, **kwargs):
...
+        if self._device_arg_id is not None:
+            device_arg_id: int = self._device_arg_id  # silent linter error
+
+            def backend_select(*args, **kwargs):
+                device = args[device_arg_id].type
+                if device not in self._backend_fns:
+                    raise RuntimeError(
+                        f"{self._name} does not have a kernel registered for {device}"
+                    )
+                return self._backend_fns[device](*args, **kwargs)
```
Comment (zou3519) [inline torch/_library/custom_ops.py:544]: for this to be faithful to how built-in PyTorch operators work, we should have with_keyset=True and have this do `opoverload.redispatch(keyset, *args, **kwargs)`

File: torch/_library/infer_schema.py [hunk 2]
```diff
@@ -223,3 +223,23 @@ def tuple_to_list(tuple_type: typing.Type[typing.Tuple]) -> typing.Type[typing.L
...
+
+
+def get_device_arg_id(schema: str) -> Union[int, None]:
+    """
+    Given a schema string, returns the id of the `device: torch.device` argument.
+    If it does not exist, returns None.
+    """
+    inputs = schema.split("->")[0].split("(")[1].split(")")[0].split(",")
+
+    return next((i for i, s in enumerate(inputs) if "Device device" == s.strip()), None)
```
Comment (zou3519) [inline torch/_library/infer_schema.py:245]: parsing the schema string and doing string match is a bit sketchy. Would it be possible to check the parsed schema (opoverload._schema) instead?

File: torch/_library/custom_ops.py (hunk 1 above)
Comment (zou3519) [inline torch/_library/custom_ops.py:544]: for this to be faithful to how built-in PyTorch operators work, we should have with_keyset=True and have this do `opoverload.redispatch(keyset, *args, **kwargs)`

File: test/test_custom_ops.py [hunk 3]
```diff
@@ -2969,6 +2969,26 @@ def numpy_sin_inplace(x: Tensor) -> Tensor:
...
+    def test_factory_function(self):
+        @torch.library.custom_op(
+            "_torch_testing::f", mutates_args={}, device_types="cpu"
+        )
+        def f(device: torch.device) -> Tensor:
+            return torch.ones(3)
+
+        result = f(device="cpu")
+        self.assertEqual(result.device, torch.device("cpu"))
+        self.assertEqual(result, torch.ones(3))
```
Comment (zou3519) [inline test/test_custom_ops.py:2981]: Can you test the case where we split registrations too?

```
@torch.library.custom_op(
    "_torch_testing::f", mutates_args={}, device_types="cpu"
 )
 def f() -> Tensor:
     raise NotImplementedError("NYI")

@f.register_kernel("cpu")
def _():
    return torch.zeros(3)

result = f(x)
```

This case should error out somewhere saying that we need a device: Device keyword arg

File: torch/_library/infer_schema.py (hunk 2 above)
Comment (zou3519) [reply torch/_library/infer_schema.py:245]: These helpers (at least the first one) may be useful to put into torch._library.utils

**Review by zou3519 — COMMENTED** [review]
- manual parsing of the schema string can be brittle; we should try to use the post-parsed schema (opoverload._schema)

- to be faithful to how built-in ops work, inside the BackendSelect kernel we want to call opoverload.redispatch, which will call one of the backend dispatchkeys (like DispatchKey::CPU).

File: torch/_library/infer_schema.py (hunk 2 above)
Comment (yushangdi) [reply torch/_library/infer_schema.py]: now using parsed schema, and moved helpers to torch._library.utils.

File: torch/_library/custom_ops.py (hunk 1 above)
Comment (yushangdi) [reply torch/_library/custom_ops.py]: changed.

File: torch/_library/custom_ops.py [hunk 4]
```diff
@@ -502,6 +517,43 @@ def adinplaceorview_impl(keyset, *args, **kwargs):
...
+        if not has_tensor_arg(schema) and backend_select:
+            device_arg_id = get_device_arg_id(schema)
+            if device_arg_id is None:
+                raise ValueError(
+                    "Functions without tensor inputs are required to have a `device: torch.device` argument"
+                )
+
+        if device_arg_id is not None:
+
+            def get_dispatch_key(device: str):
```
Comment (yushangdi) [inline torch/_library/custom_ops.py:532]: @zou3519 I'm wondering if there's a better way to get the dispatch key from a device string?

File: test/test_custom_ops.py (hunk 3 above)
Comment (zou3519) [reply test/test_custom_ops.py:2981]: Sorry, I meant the following:
```
@torch.library.custom_op(
    "_torch_testing::f", mutates_args={}
 )
 def f() -> Tensor:
     raise NotImplementedError("NYI")

@f.register_kernel("cpu")
def _():
    return torch.zeros(3)

result = f(x)
```

It should be fine to register a factory function that doesn't have a `device: Device` arg if there is no separate backend dispatch. So, the following should be OK:

```
@torch.library.custom_op(
    "_torch_testing::f", mutates_args={}
 )
 def f() -> Tensor:
     return torch.ones(3)
```

File: torch/_library/custom_ops.py (hunk 4 above)
Comment (zou3519) [reply torch/_library/custom_ops.py:532]: try this: https://github.com/pytorch/pytorch/blob/64a04d2225800c137c1db562877ab43cd5d76f66/torch/_library/custom_ops.py#L266

File: test/test_custom_ops.py (hunk 3 above)
Comment (yushangdi) [reply test/test_custom_ops.py:2981]: makes sense. fixed now.

File: torch/_library/utils.py [hunk 5]
```diff
@@ -256,3 +256,25 @@ def has_kwarg_only_tensors(schema: _C.FunctionSchema):
...
+    """
+    return any(
+        (is_tensor_like_type(a.type) or is_tensorlist_like_type(a.type))
+        for a in schema.arguments
+    )
+
+
+def get_device_arg_id(schema: _C.FunctionSchema) -> Union[int, None]:
+    """
+    Given a schema, returns the id of the `device: torch.device` argument.
```
Comment (zou3519) [inline torch/_library/utils.py:274]: nit: id -> index

**Review by zou3519 — APPROVED** [review]
LGTM, some minor comments around testing and naming, please read

File: test/test_custom_ops.py [hunk 6]
```diff
@@ -2969,6 +2969,48 @@ def numpy_sin_inplace(x: Tensor) -> Tensor:
...
+            ValueError,
+            "Functions without tensor inputs are required to have a `device: torch.device` argument",
+        ):
+
+            @f3.register_kernel("cpu")
+            def _():
+                return torch.zeros(3)
+
+            result = f(x)
+
```
[the remark this answers is not in the document]
Comment (yushangdi) [reply test/test_custom_ops.py:3013]: Do you mean the test below? (with device argument in the register_kernel function). I added the test below.

```python
        @torch.library.custom_op("_torch_testing::f4", mutates_args={})
        def f4(device: torch.device) -> Tensor:
            raise NotImplementedError("NYI")

        @f4.register_kernel("cpu")
        def _(device: torch.device):
            return torch.zeros(3)

        result = f(device="cpu")
        self.assertEqual(result.device, torch.device("cpu"))
        self.assertEqual(result, torch.ones(3))

```

[the remark this answers is not in the document]
Comment (zou3519) [reply test/test_custom_ops.py:3013]: yes

**Pull request closed without merging**